src/analyses/variable-sensitivity/liveness_context.cpp

```diff
@@ -0,0 +1,220 @@
+}
+
+abstract_object_pointert liveness_contextt::combine(
+  const region_context_ptrt &other,
+  combine_fn fn) const
+{
+  auto combined_child = fn(child_abstract_object, other->child_abstract_object);
+  auto location_match = get_location() == other->get_location();
+
+  if(combined_child.modified || location_match)
+  {
+    const auto &result =
+      std::dynamic_pointer_cast<liveness_contextt>(mutable_clone());
+    result->set_child(combined_child.object);
+    result->reset_location();
+    return result;
+  }
+
+  return shared_from_this();
+}
+
+void liveness_contextt::reset_location()
+{
+  assign_location.reset();
+}
+
+context_abstract_objectt::context_abstract_object_ptrt
+liveness_contextt::update_location_context_internal(
+  const locationst &locations) const
+{
+  auto result = std::dynamic_pointer_cast<liveness_contextt>(mutable_clone());
+  result->set_location(*locations.cbegin());
+  return result;
+}
+
+void liveness_contextt::set_location(const locationt &location)
+{
+  assign_location.emplace(location);
+}
+
+/**
+ * Output a representation of the value of this abstract object
+ *
+ * \param out the stream to write to
+ * \param ai the abstract interpreter that contains the abstract domain
+ * (that contains the object ... )
+ * \param ns the current namespace
+ */
+void liveness_contextt::output(
+  std::ostream &out,
+  const ai_baset &ai,
+  const namespacet &ns) const
+{
+  context_abstract_objectt::output(out, ai, ns);
+
+  if(assign_location.has_value())
+    out << " @ [" << assign_location.value()->location_number << "]";
+  else
+    out << " @ [undefined]";
+}
+
+/**
+ * Determine whether 'this' abstract_object has been modified in comparison
+ * to a previous 'before' state.
+ *
+ * \param before the abstract_object_pointert to use as a reference to
+ * compare against
+ *
+ * \return true if 'this' is considered to have been modified in comparison
+ * to 'before', false otherwise.
+ */
+bool liveness_contextt::has_been_modified(
+  const abstract_object_pointert &before) const
+{
+  if(this == before.get())
+    return false;
+
+  auto before_context =
+    std::dynamic_pointer_cast<const liveness_contextt>(before);
+
+  if(!before_context)
+  {
+    // The other context is not something we understand, so must assume
+    // that the abstract_object has been modified
+    return true;
+  }
+
+  // Even if the pointers are different, it maybe that it has changed only
+  // because of a merge operation, rather than an actual write. Given that
+  // this class has knowledge of where writes have occured, use that
+  // information to determine if any writes have happened and use that as the
+  // proxy for whether the value has changed or not.
+  //
+  // For two sets of last written locations to match,
+  // each location in one set must be equal to precisely one location
+  // in the other, since a set can assume at most one match
+  return before_context->get_location() != get_location();
```

Comment:
As `locationt`'s are iterators into (potentially different) lists, we should compare location numbers instead of directly comparing locations.